qiita_db/sql_connection.py

```diff
@@ -630,3 +632,381 @@ def get_temp_queue(self):
+                sql, sql_args = self._replace_placeholders(sql, sql_args)
+
+                # Execute the current SQL command
+                try:
+                    cur.execute(sql, sql_args)
+                except Exception as e:
+                    # We catch any exception as we want to make sure that we
+                    # rollback every time that something went wrong
+                    self._raise_execution_error(sql, sql_args, e)
+
+                try:
+                    res = cur.fetchall()
+                except ProgrammingError as e:
+                    # At this execution point, we don't know if the sql query
+                    # that we executed should retrieve values from the database
+                    # If the query was not supposed to retrieve any value
+                    # (e.g. an INSERT without a RETURNING clause), it will
+                    # raise a ProgrammingError. Otherwise it will just return
+                    # an empty list
+                    res = None
+                except PostgresError as e:
+                    # Some other error happened during the execution of the
+                    # query, so we need to rollback
+                    self._raise_execution_error(sql, sql_args, e)
+
+                # Store the results of the current query
+                self._results.append(res)
+
+        # wipe out the already executed queries
+        self._queries = []
+
+        return self._results
+
+    @_checker
+    def execute(self):
+        """Executes the transaction
+
+        Returns
+        -------
+        list of DictCursor
+            The results of all the SQL queries in the transaction
+
+        Raises
+        ------
+        RuntimeError
+            If invoked outside a context
+
+        Notes
+        -----
+        If any exception occurs during the execution transaction, a rollback
+        is executed an no changes are reflected in the database.
+        When calling execute, the transaction will never be committed, it will
+        be automatically committed when leaving the context
+
+        See Also
+        --------
+        execute_fetchlast
+        """
+        try:
+            return self._execute()
+        except Exception:
+            self.rollback()
+            raise
+
+    @_checker
+    def execute_fetchlast(self):
+        """Executes the transaction and returns the last result
+
+        This is a convenient function that is equivalent to
+        `self.execute()[-1][0][0]`
+
+        Returns
+        -------
+        object
+            The first value of the last SQL query executed
+
+        See Also
+        --------
+        execute
+        """
+        return self.execute()[-1][0][0]
+
+    @_checker
+    def commit(self):
+        """Commits the transaction and reset the queries
+        Raises
+        ------
+        RuntimeError
+            If invoked outside a context
+        """
+        self._connection.commit()
```

Comment:
You're probably right, although if those calls fail, the transaction will not be left IDLE, so it is blocking the DB... If these calls fail we have a bigger problem that I don't know how to solve, since the way of resetting the transaction is doing a rollback or a commit. A potential solution will be forcing to close the connection as it internally will do a rollback, but not sure if in that such-a-hard-case-to-test this will work. Worst case scenario, these exceptions are propagated and I don't think they will be caught anywhere, which I think is a desirable behavior if this case arises.